utils/binutils.ml

```diff
@@ -0,0 +1,577 @@
+    in
+    let endianness =
+      match magic, Sys.big_endian with
+      | (MH_MAGIC | MH_MAGIC_64), false
+      | (MH_CIGAM | MH_CIGAM_64), true -> LE
+      | (MH_MAGIC | MH_MAGIC_64), true
+      | (MH_CIGAM | MH_CIGAM_64), false -> BE
+    in
+    let d = {ic; endianness; bitness} in
+    let header = read_header d in
+    let load_commands = read_load_commands d header in
+    let symbols = read_symbols d load_commands in
+    let symbol_offset = symbol_offset symbols in
+    let defines_symbol = defines_symbol symbols in
+    {symbol_offset; defines_symbol}
+end
+
+module FlexDLL = struct
+
+  let header_size = 24
+
+  type header =
+    {
+      e_lfanew: int64;
+      number_of_sections: int;
+      size_of_optional_header: int;
+      characteristics: int;
+    }
+
+  let read_header e_lfanew d buf =
+    let number_of_sections = get_uint16 d buf 6 in
+    let size_of_optional_header = get_uint16 d buf 20 in
+    let characteristics = get_uint16 d buf 22 in
+    {e_lfanew; number_of_sections; size_of_optional_header; characteristics}
+
+  type optional_header_magic =
+    | PE32
+    | PE32PLUS
+
+  type optional_header =
+    {
+      magic: optional_header_magic;
+      image_base: int64;
+    }
+
+  let read_optional_header d {e_lfanew; size_of_optional_header; _} =
+    let buf = load_bytes d Int64.(add e_lfanew (of_int header_size)) size_of_optional_header in
+    let magic =
+      match get_uint16 d buf 0 with
+      | 0x10b -> PE32
+      | 0x20b -> PE32PLUS
+      | n -> raise (Error (Unsupported ("optional header magic", n)))
+    in
+    let image_base =
+      match magic with
+      | PE32 -> unsigned_of_int32 (get_uint32 d buf 28)
+      | PE32PLUS -> get_uint64 d buf 24
+    in
+    {magic; image_base}
+
+  type section =
+    {
+      name: string;
+      virtual_size: int;
+      virtual_address: int64;
+      size_of_raw_data: int;
+      pointer_to_raw_data: int64;
+    }
+
+  let section_header_size = 40
+
+  let read_sections d {e_lfanew; number_of_sections; size_of_optional_header; _} =
+    let buf =
+      load_bytes d
+        Int64.(add e_lfanew (of_int (header_size + size_of_optional_header)))
+        (number_of_sections * section_header_size)
+    in
+    let mk i =
+      let base = i * section_header_size in
+      let name = name_at ~max_len:8 buf (base + 0) in
+      let virtual_size = get_uint d buf (base + 8) in
+      let virtual_address = unsigned_of_int32 (get_uint32 d buf (base + 12)) in
+      let size_of_raw_data = get_uint d buf (base + 16) in
+      let pointer_to_raw_data = unsigned_of_int32 (get_uint32 d buf (base + 20)) in
+      {name; virtual_size; virtual_address; size_of_raw_data; pointer_to_raw_data}
+    in
+    Array.init number_of_sections mk
+
+  type symbol =
+    {
+      name: string;
+      address: int64;
+    }
+
+  let load_section_body d {size_of_raw_data; pointer_to_raw_data; _} =
+    load_bytes d pointer_to_raw_data size_of_raw_data
```

Comment:
It was not obvious to me from the docs that `pointer_to_raw_data` didn't need the `e_lfanew` offset addition. But if it works I guess that's ok :-)

Reply:
Yes, it was trial-and-error mostly.